Feat/similarity table

melisande1 opened the pull request.

pep8speaks commented:
Hello @melisandequentzel! Thanks for opening this PR. 

* In the file [`credentialdigger/client.py`](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/credentialdigger/client.py):

> [Line 470:20](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/credentialdigger/client.py#L470): [F821](https://duckduckgo.com/?q=pep8%20F821) undefined name 'Error'
> [Line 475:20](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/credentialdigger/client.py#L475): [F821](https://duckduckgo.com/?q=pep8%20F821) undefined name 'Error'



* In the file [`credentialdigger/client_sqlite.py`](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/credentialdigger/client_sqlite.py):

> [Line 1:1](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/credentialdigger/client_sqlite.py#L1): [F401](https://duckduckgo.com/?q=pep8%20F401) 're' imported but unused
> [Line 182:1](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/credentialdigger/client_sqlite.py#L182): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 208:80](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/credentialdigger/client_sqlite.py#L208): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 226:80](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/credentialdigger/client_sqlite.py#L226): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 229:80](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/credentialdigger/client_sqlite.py#L229): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (94 > 79 characters)
> [Line 234:80](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/credentialdigger/client_sqlite.py#L234): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 79 characters)
> [Line 458:1](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/credentialdigger/client_sqlite.py#L458): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 461:66](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/credentialdigger/client_sqlite.py#L461): [E703](https://duckduckgo.com/?q=pep8%20E703) statement ends with a semicolon
> [Line 463:71](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/credentialdigger/client_sqlite.py#L463): [E703](https://duckduckgo.com/?q=pep8%20E703) statement ends with a semicolon
> [Line 464:80](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/credentialdigger/client_sqlite.py#L464): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (93 > 79 characters)
> [Line 560:1](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/credentialdigger/client_sqlite.py#L560): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 592:80](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/credentialdigger/client_sqlite.py#L592): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (102 > 79 characters)
> [Line 593:80](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/credentialdigger/client_sqlite.py#L593): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 596:80](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/credentialdigger/client_sqlite.py#L596): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 598:80](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/credentialdigger/client_sqlite.py#L598): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (93 > 79 characters)
> [Line 600:80](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/credentialdigger/client_sqlite.py#L600): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 605:1](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/credentialdigger/client_sqlite.py#L605): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file



* In the file [`ui/backend/client_ui.py`](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/ui/backend/client_ui.py):

> [Line 104:1](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/ui/backend/client_ui.py#L104): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

* In the file [`ui/backend/client_ui_sqlite.py`](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/ui/backend/client_ui_sqlite.py):

> [Line 1:1](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/ui/backend/client_ui_sqlite.py#L1): [F401](https://duckduckgo.com/?q=pep8%20F401) 're' imported but unused
> [Line 187:80](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/ui/backend/client_ui_sqlite.py#L187): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 190:80](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/ui/backend/client_ui_sqlite.py#L190): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (94 > 79 characters)
> [Line 193:16](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/ui/backend/client_ui_sqlite.py#L193): [F821](https://duckduckgo.com/?q=pep8%20F821) undefined name 'Error'
> [Line 195:80](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/ui/backend/client_ui_sqlite.py#L195): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 79 characters)
> [Line 198:5](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/ui/backend/client_ui_sqlite.py#L198): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 229:80](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/ui/backend/client_ui_sqlite.py#L229): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (102 > 79 characters)
> [Line 230:80](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/ui/backend/client_ui_sqlite.py#L230): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 233:80](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/ui/backend/client_ui_sqlite.py#L233): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 235:80](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/ui/backend/client_ui_sqlite.py#L235): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (93 > 79 characters)
> [Line 237:80](https://github.com/SAP/credential-digger/blob/4dd5136b2a8449ee7c9e19171af119aa44f32fbc/ui/backend/client_ui_sqlite.py#L237): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)



Do see the [Hitchhiker's guide to code style](https://goo.gl/hqbW4r)

marcorosa commented on credentialdigger/client.py:
```diff
@@ -453,6 +484,17 @@ def get_discovery_group(self, query, state_query, repo_url, state=None):
             cursor.execute(query, (repo_url,))
         return cursor.fetchall()
 
+    def get_embedding(self, query, discovery_id=None, snippet=None):
```
missing documentation

marcorosa commented on credentialdigger/client.py:
```diff
@@ -453,6 +484,17 @@ def get_discovery_group(self, query, state_query, repo_url, state=None):
+        try:
+            if discovery_id:
+                cursor.execute(query, (discovery_id,))
+            else:
```
What happens if both the `discovery_id` and the `snippet` are `None`?

marcorosa commented on credentialdigger/client.py:
```diff
@@ -559,7 +601,7 @@ def update_discovery_group(self, query, new_state, repo_url, file_name=None,
     def scan(self, repo_url, category=None, models=None, exclude=None,
              force=False, debug=False, generate_snippet_extractor=False,
-             local_repo=False, git_token=None):
+             similarity=False, local_repo=False, git_token=None):
```
if you want to add a new parameter, then you also have to provide its documentation in the docstrings (of this method and the others where you pass it)

marcorosa commented on credentialdigger/client.py:
```diff
@@ -874,6 +917,8 @@ def _scan(self, repo_url, scanner, models=None, force=False, debug=False,
             except ModuleNotFoundError:
                 logger.warning('SnippetModel not found. Skip it.')
 
+        if similarity:
```
if it takes some time, please add a log. Something like
`logger.info('Build embedding model for this repository')`
otherwise the scan will look like pending

marcorosa commented on credentialdigger/client.py:
```diff
@@ -1017,7 +1066,6 @@ def update_similar_snippets(self,
                                 threshold=0.96):
         """ Find snippets that are similar to the target
         snippet and update their state.
-
```
why have you removed the blank lines from this docstring?

marcorosa commented on credentialdigger/client.py:
```diff
@@ -1031,28 +1079,29 @@ def update_similar_snippets(self,
             The number of similar snippets found and updated
         """
 
         discoveries = self.get_discoveries(repo_url, file_name)
```
do you need all the discoveries or only the "new" ones?

marcorosa commented on credentialdigger/client.py:
```diff
@@ -1031,28 +1079,29 @@ def update_similar_snippets(self,
-                    n_updated_snippets += 1
-                    self.update_discovery(d['id'], state)
-        return n_updated_snippets
+        if target_embedding:
```
to make this nested block more readable and easier:

```
# comment to explain it
if not target_embedding:
    return 0

# We have a target embedding (or another comment)
for d in discoveries:
    ...
```

marcorosa commented on credentialdigger/client.py:
```diff
@@ -1031,28 +1079,29 @@ def update_similar_snippets(self,
+                    d['state'] != state
+                    and self.get_embedding(discovery_id=d['id'])
+                ):
+                    # Compute similarity of target snippet and snippet
```
this comment is a little confusing: didn't we calculate the embedding of the target snippet before? 
Indeed, here you're using `target_embedding` and not `target_snippet`

marcorosa commented on credentialdigger/client.py:
```diff
@@ -1031,28 +1079,29 @@ def update_similar_snippets(self,
+                    similarity = compute_similarity(target_embedding,
+                                                    embedding)
+                    if similarity > threshold:
+                        n_updated_snippets += 1
+                        self.update_discovery(d['id'], state)
```
invert these 2 lines: before you update the discovery and then you update `n`.
It doesn't change anything, but it seems more logical

marcorosa commented on credentialdigger/client_postgres.py:
```diff
@@ -195,6 +198,66 @@ def add_rule(self, regex, category, description=''):
                     VALUES (%s, %s, %s) RETURNING id')
 
+    def add_embedding(self, discovery_id, embedding=None, repo_url=''):
+        """ Add an embedding to the embeddings table.
```
missing blank line before Parameters

marcorosa commented on credentialdigger/client_postgres.py:
```diff
@@ -195,6 +198,66 @@ def add_rule(self, regex, category, description=''):
+
+        cursor = self.db.cursor()
+        snippet = self.get_discovery(discovery_id)['snippet']
+        if embedding is None:
```
`if not embedding`:

marcorosa commented on credentialdigger/client_postgres.py:
```diff
@@ -195,6 +198,66 @@ def add_rule(self, regex, category, description=''):
+            self.db.rollback()
+
+    def add_embeddings(self, repo_url):
+        """ Bulk add embeddings.
```
missing blank line

marcorosa commented on credentialdigger/client_postgres.py:
```diff
@@ -195,6 +198,66 @@ def add_rule(self, regex, category, description=''):
+        Parameters
+        ----------
+        repo_url: str
+            The discoveries' reposiroty url
```
typo

marcorosa replied:
@melisande1 you didn't notice this

marcorosa commented on credentialdigger/client_postgres.py:
```diff
@@ -249,6 +313,46 @@ def delete_discoveries(self, repo_url):
 
+    def delete_embedding(self, discovery_id):
+        """ Delete an embedding.
+        Parameters
```
missing blank line

marcorosa commented on credentialdigger/client_postgres.py:
```diff
@@ -195,6 +198,66 @@ def add_rule(self, regex, category, description=''):
+            query = 'INSERT INTO embeddings (id, embedding, snippet, repo_url) \
+                    VALUES (%s, %s, %s, %s);'
+            insert_tuples = []
+            for i in range(len(discoveries_ids)):
```
`for current_disc in discoveries:` ? Unless you need `i`

marcorosa commented on credentialdigger/client_postgres.py:
```diff
@@ -249,6 +313,46 @@ def delete_discoveries(self, repo_url):
+
+    def delete_embeddings(self, repo_url):
+        """ Delete all embeddings from a repository.
+        Parameters
```
same comment

marcorosa commented on sql/create_table.sql:
```diff
@@ -28,3 +28,11 @@ CREATE TABLE discoveries (
+CREATE TABLE embeddings (
+  id INTEGER REFERENCES discoveries,
+  embedding FLOAT [],
+  snippet TEXT DEFAULT '',
+  repo_url TEXT,
```
don't they reference discoveries.snippet and repos.url?

marcorosa commented on credentialdigger/snippet_similarity.py:
```diff
@@ -53,38 +53,41 @@ def compute_snippet_embedding(snippet, model):
     ----------
-    emb_1: np.array
+    emb_1: list
         first snippet embedding
```
Initial letter should be capitalised (also for `emb_2`)

marcorosa commented on credentialdigger/snippet_similarity.py:
```diff
@@ -53,38 +53,41 @@ def compute_snippet_embedding(snippet, model):
+    return snippet_embedding.tolist()
 
 
 def compute_similarity(emb_1, emb_2):
```
please use better variables names (e.g., embedding_1 and 2), otherwise we risk not to understand it in 6 months :D

marcorosa commented on ui/res/js/discoveries.js:
```diff
@@ -145,14 +145,18 @@ function initDiscoveriesDataTable() {
             </tbody>
           </table><div>`;
+          
+          const actions = discoveriesBtnGroupTemplate('Mark as')+`
```
actions is not a very good name. We risk to have a conflict because it's too general (and maybe prone to be used for something else)

marcorosa commented on ui/server.py:
```diff
@@ -277,7 +277,8 @@ def scan_repo():
         'git_token': git_token,
-        'local_repo': local_repo
+        'local_repo': local_repo,
+        'similarity': True
```
should we leave it mandatory or ask the user?

marcorosa commented on ui/server.py:
```diff
@@ -407,6 +408,22 @@ def update_discovery_group():
+    response2 = c.update_similar_snippets(target_snippet,
+                                          state,
+                                          url)
+    if (response1 is False or response2 is False):
```
( ) not needed

marcorosa commented on ui/backend/client_ui_sqlite.py:
```diff
@@ -165,3 +170,33 @@ def get_files_summary(self, repo_url):
                 " GROUP BY file_name"
             ))
+
+    def add_embeddings(self, repo_url):
```
missing documentation

marcorosa commented on ui/backend/client_ui_sqlite.py:
```diff
@@ -165,3 +170,33 @@ def get_files_summary(self, repo_url):
+            cursor.executemany(query, insert_tuples)
+            self.db.commit()
+        except Error:
+            self.db.rollback()
```
what are you doing in this except block?

marcorosa commented on ui/backend/client_ui_sqlite.py:
```diff
@@ -165,3 +170,33 @@ def get_files_summary(self, repo_url):
+        snippets = [d['snippet'] for d in discoveries]
+        model = build_embedding_model()
+        embeddings = [compute_snippet_embedding(s, model) for s in snippets]
+        embedding_strings = ["" for i in discoveries]
+        for i in range(len(discoveries_ids)):
+            for emb in embeddings[i]:
+                embedding_strings[i] += str(emb) + ","
```
not very elegant. Isn't there another way to do it?

marcorosa reviewed (changes requested):
I have several comments (you can follow the review) that are "minor", but I also have a major one:
The add/get/delete embedding methods are almost the same both for the sqlite and postgres classes. So, can't we move them to the parent class like we did for other methods (e.g., add_discovery)?
In the end, the process is the same, and the only difference is in the query format.
This would considerably help the maintainability of the project on the long term.

melisande1 commented:
Thank you for the review, will make the requested changes and will reply to specific comments.
Indeed, I will move parts of the code to the parent class in get/add/delete_embeddings, it makes more sense.

melisande1 reviewed (commented):
I replied to some comments, will make the changes accordingly (and will make changes where comments didn't require a reply)

melisande1 commented:
@marcorosa I made all the requested changes (with the exception of adding the references in the embeddings table, need to figure out why it doesn't work)

marcorosa commented on credentialdigger/client_postgres.py:
```diff
@@ -373,6 +457,37 @@ def get_discovery_group(self, repo_url, state=None):
+
+        Parameters
+        ----------
+        discovery_id: int
+            The id of the discovery whose embedding is being retrieved
+        snippet: str
```
both of them are optional arguments

marcorosa commented on credentialdigger/client_postgres.py:
```diff
@@ -373,6 +457,37 @@ def get_discovery_group(self, repo_url, state=None):
+        Returns
+        -------
+        list
+            The embedding for the provided
```
a list of embeddings (floats) ?

melisande1 replied:
yes I'm not sure how to specify the list type in the docstring

marcorosa commented on credentialdigger/client_sqlite.py:
```diff
@@ -181,6 +192,60 @@ def add_discoveries(self, discoveries, repo_url):
                 state=d['state']
             ), discoveries)
 
+    def add_embedding(self, discovery_id, embedding=None, repo_url=''):
```
I assume this method (and add_embeddings) can be moved (maybe partially, maybe wholly) in the parent class

melisande1 replied:
It's already largely in the parent class:
      return super().add_embedding(query,
                                     discovery_id,
                                     snippet,
                                     embedding_str,
                                     repo_url)

marcorosa commented on credentialdigger/client_sqlite.py:
```diff
@@ -398,6 +499,38 @@ def get_discovery_group(self, repo_url, state=None):
+        embedding = []
+        if str_embedding:
+            embedding = [float(emb)
+                         for emb in str_embedding[0].split(",")[:-1]]
```
didn't you use ; instead of ,?

melisande1 replied:
no, I used ,

marcorosa commented on credentialdigger/client_sqlite.py:
```diff
@@ -398,6 +499,38 @@ def get_discovery_group(self, repo_url, state=None):
+
+        Returns
+        ------
+        list
+            The embedding
```
ok I understood it later

marcorosa commented on credentialdigger/client_sqlite.py:
```diff
@@ -398,6 +499,38 @@ def get_discovery_group(self, repo_url, state=None):
+        embedding = []
+        if str_embedding:
+            embedding = [float(emb)
+                         for emb in str_embedding[0].split(",")[:-1]]
```
wasn't `str_embedding` a string?
why `str_embedding[0].split(...)`? it will split 1 letter

melisande1 replied:
Actually there is a mistake in the docstring of get_embedding: str_embedding is a tuple containing a single string, not a string

melisande1 replied:
I will move the [0] to the parent definition, so that get_embedding does indeed return a string/list, but as is, the code works

marcorosa commented on ui/backend/client_ui.py:
```diff
@@ -115,12 +115,12 @@ def update_similar_snippets(self,
         state: str
-            state to update similar snippets to
+            State to update similar snippets to
         repo_url: str
```
missing description (even if it's trivial)

marcorosa commented on ui/backend/client_ui.py:
```diff
@@ -115,12 +115,12 @@ def update_similar_snippets(self,
         ----------
         target_snippet: str
```
missing (trivial) description

marcorosa commented on ui/backend/client_ui.py:
```diff
@@ -131,19 +131,23 @@ def update_similar_snippets(self,
-        model = build_embedding_model()
-        target_snippet_embedding = compute_snippet_embedding(target_snippet,
-                                                             model)
+        discoveries = self.get_discoveries(repo_url, file_name)[1]
```
this method returns a list of dictionaries. Why are you taking [1]? if you don't have discoveries (or if you have only 1) this should raise a IndexError

melisande1 replied:
No, in the UI classes, this method returns the number of discoveries as 1st element and the actual discoveries as 2nd element

marcorosa commented on ui/backend/client_ui.py:
```diff
@@ -131,19 +131,23 @@ def update_similar_snippets(self,
-                n_updated_snippets += 1
-                self.update_discovery(d['id'], state)
-        return n_updated_snippets
+        if not target_embedding:
+            return 0
+        else:
+            for d in discoveries:
+                if (
+                    d['state'] != state
+                    and self.get_embedding(discovery_id=d['id'])
+                ):
+                    # Compute similarity of target embedding and embedding
+                    embedding = self.get_embedding(discovery_id=d['id'])
+                    similarity = compute_similarity(target_embedding,
+                                                    embedding)
+                    if similarity > threshold:
+                        n_updated_snippets += 1
+                        self.update_discovery(d['id'], state)
+            return n_updated_snippets
```
isn't this block of code same as in the parent class?

melisande1 replied:
yes this block is the same as in the parent class, but unlike the entire method, so I cannot simply delete the method from the UI class

marcorosa commented on ui/backend/client_ui_postgres.py:
```diff
@@ -1,5 +1,10 @@
+from sqlite3 import Error
```
sqlite's Error in Postgres? Are you sure?

marcorosa commented on ui/backend/client_ui_postgres.py:
```diff
@@ -162,3 +167,33 @@ def get_files_summary(self, repo_url):
                 " GROUP BY file_name"
             ))
+
+    def add_embeddings(self, repo_url):
```
again copy-pasted code from a parent class

melisande1 replied:
not exactly copy pasted, it's the same issue as update_similar_discoveries, there's the get_discoveries[1], that I can't get rid of

marcorosa commented on ui/server.py:
```diff
@@ -407,6 +408,22 @@ def update_discovery_group():
+    response2 = c.update_similar_snippets(target_snippet,
+                                          state,
+                                          url)
+    if response1 is False or response2 is False:
```
`if not response1 or not response2:`
(more clear)

melisande1 replied:
I agree, but in the update_discovery_group, there `if response is False`

marcorosa commented on ui/server.py:
```diff
@@ -407,6 +408,22 @@ def update_discovery_group():
+                                          state,
+                                          url)
+    if response1 is False or response2 is False:
+        return 'Error in updating similar snippets', 500
```
can this happen? 
According to the `if`, you may update the similar ones without updating the "discovery group"

marcorosa replied:
shouldn't we do something like
```
response1 = c.update_discovery_group(state, url, file, target_snippet)

if not response1:
    return '...', 500
response2 = c.update_similar_snippets(target_snippet,state,url)
if not response2:
    return '...', 500
return '...', 200
```

melisande1 replied:
yes it's better, I thought about it

marcorosa reviewed (changes requested):
Hi, thanks for the wide PR!

I have some minor comments on style and structures. I think they will be easy to address.

After that, we can discuss on 2 concerns that I have.
1- At first I didn't understand why you had to use a string for embeddings in sqlite and a list of floats in postgres, then I got it. I wonder if this approach is better than using serialisation (i.e., `pickle`).
2- I'm pretty sure that we can move most/all the `add_embedding(s)` methods to Client, since they are all copy-pasted. Maybe there is some post-processing due in the children, but anyway we have to sort this out after the point at 1)

marcorosa commented:
By the way I don't know why @pep8speaks didn't send any message. Either you have been perfect with the style rules, or we'll receive its alert later once I'll open the PR to the main branch

melisande1 commented:
Hi, thank you for the review! I will make all the changes. Pep did speak, its message is hidden among the commits, when you expand them you can see it: the style is fine. I have two remarks, one regarding moving the add_embeddings from the UI child classes to the regular Client: there is a difference, it's in the get_discoveries[1] vs get_discoveries, I don't think there's a way around, unless passing the discoveries as arguments to the update_discoveries method, but it will likely make things more complicated. Regarding the other tiny remark I have, I'm replying to the specific comment.

melisande1 commented on credentialdigger/client.py:
```diff
@@ -112,12 +113,78 @@ def add_discovery(self, query, file_name, commit_id, line_number, snippet,
+            self.db.commit()
+        except self.Error:
+            self.db.rollback()
+            map(lambda disc_id, emb: self.add_embedding(disc_id,
+                                                        emb,
+                                                        repo_url=repo_url),
+                zip(discoveries_ids, embeddings))
```
yes, likely, I copied the layout of add_discoveries, so I might as well, just remove the exception part 
and just do a rollback

melisande1 commented on credentialdigger/client_postgres.py:
```diff
@@ -373,6 +457,37 @@ def get_discovery_group(self, repo_url, state=None):
+        embedding = super().get_embedding(query=query,
+                                          discovery_id=discovery_id,
+                                          snippet=snippet)
+        if embedding:
+            return embedding[0]
+        else:
+            return None
```
yes I will move it

melisande1 commented on credentialdigger/client_postgres.py:
```diff
@@ -227,6 +275,7 @@ def delete_repo(self, repo_url):
         bool
             `True` if the repo was successfully deleted, `False` otherwise
         """
+        self.delete_embeddings(repo_url)
```
ok

melisande1 commented on credentialdigger/client.py:
```diff
@@ -1039,20 +1202,25 @@ def update_similar_snippets(self,
+                    and self.get_embedding(discovery_id=d['id'])
+                ):
+                    # Compute similarity of target embedding and embedding
+                    embedding = self.get_embedding(discovery_id=d['id'])
```
okay, I will make it one call

melisande1 commented on credentialdigger/client.py:
```diff
@@ -1039,20 +1202,25 @@ def update_similar_snippets(self,
+        if not target_embedding:
+            return 0
+        # We have a target embedding for the target snippet
+        else:
```
okay

melisande1 commented on credentialdigger/client.py:
```diff
@@ -883,11 +1042,15 @@ def _scan(self, repo_url, scanner, models=None, force=False, debug=False,
                     curr_d['file_name'], curr_d['commit_id'],
                     curr_d['line_number'], curr_d['snippet'], repo_url,
                     curr_d['rule_id'], curr_d['state'])
+                if similarity:
+                    self.add_embedding(new_id, repo_url)
```
this is just for the debug mode, where the if part is made of multiple calls to add_discovery, so I made it similar, this way one can precisely locate where a bug might arise. I don't think there's a significant difference in speed between add_embedding vs add_embeddings, especially since the cursor is not closed

melisande1 commented on credentialdigger/client.py:
```diff
@@ -1039,20 +1202,25 @@ def update_similar_snippets(self,
+            for d in discoveries:
+                if (
+                    d['state'] != state
+                    and self.get_embedding(discovery_id=d['id'])
```
The issue with that is that we only want to retrieve embeddings that are not from the current state (and the state in not present in the discovery table). Moreover, I didn't make a get_embeddings method, because just getting a list of embeddings without any id would not be very useful, so it would need to return a list of tuples with ids and embeddings. This matching between embeddings and ids is more easily achieved using a get_embedding with the id as argument. And when I see the similarity update being performed rather instantly, even on large repos, I don't think there's a problem in making individual calls to get_embedding, but I'm happy to try otherwise

melisande1 commented on credentialdigger/client_sqlite.py:
```diff
@@ -181,6 +192,60 @@ def add_discoveries(self, discoveries, repo_url):
+        if not embedding:
+            model = build_embedding_model()
+            embedding = compute_snippet_embedding(snippet, model)
+        embedding_str = ""
+        for emb in embedding:
+            embedding_str += str(emb) + ","
```
yes I should have used join

melisande1 commented:
@marcorosa I made the requested changes. Now I'm ready to make changes to add_embeddings and update_similar_snippets to remove parts of the code from the child UI classes, but I see no simple solution, except to check for get_discoveries's output length in parent class

melisande1 commented:
@marcorosa I moved the two methods from UI child classes to the parent Client class, by checking for output length of get_discoveries. I'm done with the changes.

melisande1 commented:
@marcorosa Hi, I made complementary changes, so as to comply with all of your comments/requirements. Namely I changed the way of retrieving embeddings in update_similar_snippets from get_embedding to get_embeddings (I added a get_embeddings method), and I created a compute_repo_embeddings method so as to refactor add_embeddings (to move additional code to the parent class)

marcorosa commented on credentialdigger/client.py:
```diff
@@ -112,12 +113,75 @@ def add_discovery(self, query, file_name, commit_id, line_number, snippet,
+                      repo_url,
+                      embedding=None):
+        """ Add an embedding to the embeddings table.
+        Parameters
```
missing blank line above

marcorosa commented on credentialdigger/client.py:
```diff
@@ -1037,22 +1251,24 @@ def update_similar_snippets(self,
-        target_snippet_embedding = compute_snippet_embedding(target_snippet,
-                                                             model)
+        disc = self.get_discoveries(repo_url, file_name)
+        discoveries = disc[1] if len(disc) == 2 else disc
```
please put a comment because this instruction is not clear

marcorosa replied:
why should this repo have only 2 discoveries?

melisande1 replied:
That's the part where I'm handling the different output types of get_discoveries in the UI vs regular clients. In the UI classes, the discoveries are given by get_discoveries()[1] (while the discoveries count is givent by get_discoveries()[0]), so checking for output length allows to retrieve the discoveries in case the client is a UI client. I don't think that's a proper way of doing it, but otherwise I need to copy paste the code in the UI classes, which you don't seem to agree with.

marcorosa replied:
Ah ok now I understand. So, if this code is specific for the UI and doesn't work with the normal Client, we have to move it to ClientUI. And here keep a method working with the calls of the Client (if really needed).

marcorosa reviewed (commented):
Hi @melisande1, thank you, great work. I will review it ASAP. 

I know sqlite has some shortcomings, the purpose behind the request to test serialization is to keep the structure of the sqlite and postgres db consistent as much as possible. So, if you can test it (without pushing your new code before I finish the review) I will very appreciate it, otherwise we can take this into consideration later. But it's an option we need to verify and address before merging into the main branch, since we are doing a major breaking change.
But again, it's just for style and consistency, I know the string hack works fine.

melisande1 commented:
Hi @marcorosa, thank you. Okay, fine I will test serialization (and will not push changes until you're done reviewing the code). I'm assuming you want to apply serialization to both sqlite and postgres, so as to keep them similar. (Initially I had in mind you wanted to make changes only to sqlite, and this would make the dbs even less similar as they are now, but I think I misunderstood)

marcorosa commented on credentialdigger/client_postgres.py:
```diff
@@ -195,6 +198,48 @@ def add_rule(self, regex, category, description=''):
             query='INSERT INTO rules (regex, category, description) \
                     VALUES (%s, %s, %s) RETURNING id')
 
+    def add_embedding(self, discovery_id, embedding=None, repo_url=''):
```
Can an embedding have an empty repo url? Isn't it an external key in the db?

melisande1 replied:
okay, I'll remove the default value

marcorosa commented on credentialdigger/client_sqlite.py:
```diff
@@ -49,6 +51,14 @@ def __init__(self, path):
 
+            CREATE TABLE IF NOT EXISTS embeddings (
+                id INTEGER REFERENCES discoveries,
+                snippet TEXT DEFAULT '',
+                embedding TEXT DEFAULT '',
```
Do we need a default=''? If I remember correctly, NULL values are allowed in the db (unless you define NOT NULL, which is not the case here)

melisande1 replied:
no I tried without, it works as well

marcorosa commented on ui/backend/client_ui.py:
```diff
@@ -101,49 +101,3 @@ def check_repo(self, repo_url, git_token=None, local_repo=False):
-
-        discoveries = self.get_discoveries(repo_url,
-                                           file_name,
-                                           state_filter='new')[1]
```
now I get why there was this 1 (follow-up from your previous comment in the parent Client class).
So, I'm wondering if we should keep this method here. Or maybe, we can delegate part of this call to Client, and keep in ClientUI the part related to finding the correct discoveries. I can help you on this.

melisande1 replied:
yes your help on the matter would be appreciated, thank you, apart from the current one, I can only think of rather complicated solutions

marcorosa reviewed (changes requested):
I have a few comments, but I can give you assistance with the `update_similar_snippets` method

marcorosa commented:
@melisande1  wrt serialization: I was thinking of it only for sqlite. I know postgres support lists (e.g., the list of floats we need for the embeddings). The lack of support is from the sqlite part, so I was thinking on it only for sqlite. But if you think we can be consistent by serializing the embeddings for both the db so to stay consistent (and possibly to have easier maintainability of the code by writing 1 single method for both the clients) then it's worthwhile to try.

melisande1 commented:
@marcorosa okay, I see, I'll try with both, then

marcorosa commented on credentialdigger/client_sqlite.py:
```diff
@@ -398,6 +493,67 @@ def get_discovery_group(self, repo_url, state=None):
+        embedding = []
+        if str_embedding:
+            embedding = [float(emb)
+                         for emb in str_embedding.split(",")]
```
please fix single ' for strings

marcorosa commented on credentialdigger/client_sqlite.py:
```diff
@@ -398,6 +493,67 @@ def get_discovery_group(self, repo_url, state=None):
+        embeddings_dict = {}
+        for id, str_embedding in str_embeddings_dict.items():
+            embeddings_dict[id] = [float(emb)
+                                   for emb in str_embedding.split(",")]
```
please fix single ' for strings

marcorosa reviewed (changes requested):
Hi, the PR now it's correct and I will merge it, but first I have one very last request: please fix a few style inconsistencies (use `'` instead of `"`), a typo, and the comments from @pep8speaks (there are some left). After these, I will merge

melisande1 commented:
Hi, thank you, I just made those changes
